checks/shell_download_validate.go

```diff
@@ -401,7 +399,71 @@ func isPipUnpinnedDownload(cmd []string) bool {
 		}
 	}
 
-	return isInstalled
+	return isInstall
+}
+
+func isPythonCommand(cmd []string) bool {
+	for _, pi := range pythonInterpreters {
+		if isBinaryName(pi, cmd[0]) {
+			return true
+		}
+	}
+	return false
+}
+
+func isUnpinnedPythonPipInstall(cmd []string) bool {
+	if !isPythonCommand(cmd) {
+		return false
+	}
+
+	isInstall := false
+	hasPip := false
+	for i := 1; i < len(cmd); i++ {
+		// Search for pip module.
+		if strings.EqualFold(cmd[i], "-m") &&
+			i < len(cmd)-1 &&
+			strings.EqualFold(cmd[i+1], "pip") {
+			hasPip = true
+			continue
+		}
+
+		if !hasPip {
+			continue
+		}
+
+		// Search for install commands.
+		if strings.EqualFold(cmd[i], "install") {
+			isInstall = true
+			continue
+		}
+
+		if !isInstall {
+			continue
+		}
+
+		// Search for `-r file`.
```

Comment:
did we add an exclusion for `pkg==ver` somewhere as well?

Reply:
`pkg==ver` is not considered pinning, is it? I'm not aware of `pkg=hash` being valid, so I consider anything as not pinned. wdut?

Reply:
`-r requirements.txt` is also not really pinning in that case either then, because it does *not* include all transitive dependencies.

The only real python solution is to use Pipfile.lock, but requires pipenv and is not supported everywhere.

Reply:
> `-r requirements.txt` is also not really pinning in that case either then, because it does _not_ include all transitive dependencies, nor does it do any kind of checksumming.

`requirements.txt` does support pinning/checksumming, see https://davidwalsh.name/hashin. I'm going to improve the check to check the content of the file, see my TODO https://github.com/ossf/scorecard/blob/main/checks/frozen_deps.go#L273

Assuming I implement the check above, `requirements.txt` will be assumed pinned in this check PR. That's why I'm assuming that anything with `-r <>` is fine.

There's a caveat to the check https://github.com/ossf/scorecard/blob/main/checks/frozen_deps.go#L273, which is that there's no guarantee the file is even called `requirements.txt`. It's a convention, but if devs have multiple files, they'll certainly rename them. Only way would be to parse the file. 

@naveensrinivasan to comment on that.

> The only real python solution is to use Pipfile.lock, but requires pipenv and is not supported everywhere.

given the above, I think requirements.txt would work.  If you think that's too many assumptions, maybe we should consider any use of `pip` as unpinned? wdut?

> 
> Pip packages are supposed to be immutable though, so a published version should never change (unless there's a bug in the infra).

we'd like to mitigate this threat with this check.

Reply:
@inferno-chromium FYI

Reply:
I see! I wasn't aware of hashin. It doesn't handle/lock transitive dependencies though right? I'm not sure if that's enough of an improvement in that case to consider this "safe".

Do others have thoughts on this?

Reply:
I'm starting to think it's brittle too. I've added a `need-discussion` label to bring this up in next meeting too.

Reply:
I had a chat w/ @rsprabery. Results:
One way to ensure the hashes are checked is thru `--require-hashes`, see https://pip.pypa.io/en/stable/cli/pip_install/#hash-checking-mode. This is simple.

Indirect dependencies are not necessarily pulled into the requirements.txt files. According to @rsprabery, one needs to use `pip-compile` to generate them from a `requirements.ini`. To verify that indirect deps are declared, we'd need to do that ourselves in scorecard by calling pypi's API or run `pip-tools` to tell us.

If we want to continue having a check for insecure uses of `pip install`, we could start by

1. Check for presence of `--require-hashes` and
2. Provide a `bootstrap-requirement.txt` containing `pip-tools` and its dependencies (about 4 of them only) along with their corresponding hash. We can rely on dependabot to bump the hashes.
2. Provide good remediation steps to explain how do use `pip-compile` properly.

We need not verify that indirect dependencies are declared for now.

For the next iteration of the check, we could update the check to:

1. Checks for indirect dependencies

What do you think?

I'll let @rsprabery comment on the above.

Reply:
I think we should just recommend `pipenv` instead, and consider all direct `pip install` commands as unpinned . It's officially supported by Python, and seems like much less effort to set up compared to all the hoops to get this working for `requirements.txt`. If we see a lot of complaints here, or false positives, perhaps we can explore this more then. Having to check for the contents of  a `requirements (or whatever).txt` file will also be difficult to do in all contexts, as it's not always going to be obvious where to find it.

We could also expand this to other Python package management solutions (e.g. Poetry) with locking supported out of the box since they are likely much easier to check (because e.g. we can almost assume pinning is done if there is a .lock file)

Reply:
Thanks everyone for the feedback. I'll consider all `pip install` as unpinned for now. We can re-visit and maybe enable it with `--require-hashes` if 1) we get push back from community OR 2) we have a good story/tooling/remediation steps. Let's keep things simple for now, we can iterate in the next version.

Reply:
@oliverchang - I didn't realize that `Pipenv` was officially supported by Python these days. Is it being bundled like `venv` and `pip` in the python3 distributions now?

In the long term we'll probably want to support `poetry`, `pip-tools`, and `Pipenv` as I don't think the Python community has settled on an approach yet, but maybe that's changed recently?

Reply:
> I think we should just recommend pipenv instead, and consider all direct pip install commands as unpinned

I think it's overkill though when all I need is to install a package in a test script. I'm planning to switch to `-r` in https://github.com/systemd/systemd/pull/21438 and I can theoretically add `--require-hashes` there since `Dependabot` supports it but switching to `pipenv` or any other tool creating virtual python environments is absolutely unwarranted in that particular case I think

Reply:
I opened https://github.com/ossf/scorecard/issues/1306 so that it won't get lost